### PR TITLE
XML protocol support for LtacProf

### PR DESCRIPTION
**Pull request by siegebell**
LtacProf currently provides the results via an info feedback, but because of the impracticality of looking at a giant dump of text, the results are pruned.

This PR adds the following commands to the XML protocol, which will provide the full profiling results in a structured manner that an IDE could allow the user to navigate.
- LtacProfReset: `unit -> unit`
- LtacProfResults: `unit -> ltacprof_treenode`

where:

``` ocaml
type ltacprof_entry = {total : float; local : float; ncalls : int; max_total : float}
type ltacprof_treenode = {entry : ltacprof_entry; children : (string * ltacprof_treenode) list }
```

### CODE REVIEW

**Comment by cpitclaudel** [comment]
+1, looks nice :) Maybe you could provide an example of input/output?

**Comment by siegebell** [comment]
Just run `coqtop -ideslave --help-XML-protocol` :)  Note that coqtop already supports "Set Ltac Profiling." and "Unset Ltac Profiling." via add().

###### Resetting the ltac profiling counters: ######
```xml
<call val="LtacProfReset"><unit/></call>
```
Example response:
```xml
<value val="good"><unit/></value>
```

###### Getting the results from ltac profiling: ######
```xml
<call val="LtacProfResults"><unit/></call>
```
Example response:
```xml
<value val="good">
  <ltacprof total="0" local="-3.0" ncalls="0" max_total="0">
    <list>
      <pair>
        <string>auto</string>
        <ltacprof total="0.1" local="0.2" ncalls="3" max_total="0.5">
          <list/>
        </ltacprof>
      </pair>
    </list>
  </ltacprof>
</value>
```

Note the the returned ltacprof results are a mirror of the data structure kept internally by LtacProf, so the the top-level ltacprof is a dummy node -- where the total time of all the tactics is given by the negation of its local time (i.e. 3.0 seconds) and the other counters are 0.

**Comment by siegebell** [comment]
Actually, I'm curious what the threshold should be for whether set/unset/reset are supported via `add(...)` versus having a dedicated xml-protocol command. The former is tied to the state and works well with backtracking, but an IDE might prefer the latter, *stateless*, way to control profiling instead.

**Comment by cpitclaudel** [comment]
Nice, thanks! WRT your question, can you comment on how this interacts with async proofs? If I set profiling and then process a bunch of proofs in parallel, will I get a profile covering all proofs? Does calling `LtacProfResults` in different states give different results? But then, what does Reset actually do?

Orthogonal to merging: Emilio (@ejgallego) is currently working on a general `Query` mechanism that would take a `state_id`, so if indeed the profile depends on the state_id then we could expose this code through that API as well.

Longer term, once we have location information, it would be nice to get line-by-line profiling information.

**Comment by siegebell** [comment]
The behavior w.r.t. async proofs is exactly what you would get when calling "Show Ltac Profile." The results depend on your current state; if you rewind, then the profile will also rewind. Now, it might make sense for LtacProfResults to accept a particular state_id to query, but such machinery is not directly/obviously exposed in ltac/profile_ltac.ml, so I do not know how to go about doing that.

**Comment by cpitclaudel** [comment]
I'm not sure I understand. What happens if you do an EditAt followed by an LtacProfResults? What about a reset at the end of the file followed by an EditAt in the middle of a proof?

**Comment by siegebell** [comment]
The details of async proofs and the interaction with EditAt are particular to LtacProf and not this PR. I do not know of any documentation of its behavior.

From my observations, LtacProf will undo the results of rewound/cancelled states when you call EditAt. I do not know whether it undoes the results of proofs that come after that point but have not been rewound/cancelled when jumping into the middle of a proof. And I'm not sure what the correct behavior should be.

**Comment by cpitclaudel** [comment]
> The details of async proofs and the interaction with EditAt are particular to LtacProf and not this PR.

Of course.

**Comment by siegebell** [comment]
Also from my observations, resetting the results and then calling EditAt will undo the effect of resetting LtacProf.

This kind of behavior makes sense for REPL. But for an IDE, I think it could be better to disallow set/unset/reset profiling commands in the script and to always report *accumulated* profiling results that are attached to no state. E.g. the counters would always increment, despite rewinding states or changing focus, until the IDE issues a reset command.

**Comment by ejgallego** [comment]
I'd like to review this patch.

I think that removing all the feedback coming from LtacProf would be good. The IDE should definitively query Coq for info and not the other way around.

**Comment by ejgallego** [comment]
LGTM! I will definitively support this in SerAPI.

**Comment by cpitclaudel** [comment]
> E.g. the counters would always increment, despite rewinding states or changing focus, until the IDE issues a reset command.

Interesting :) I'm not too sure myself, but as you pointed out this is orthogonal to your PR.

**Comment by JasonGross** [comment]
@cpitclaudel The state which is tied to state ids is the aggregated profiling results up to that point.  I don't know how the stm does rewinding.  I didn't do anything special to get async proof support, so I'm guessing that it discards the info.

> E.g. the counters would always increment, despite rewinding states or changing focus, until the IDE issues a reset command.

This was the previous behavior, before I made profiling status be synchronized with the document.  It's actually very easy to get both: just duplicate the `stack` structure using `ref` rather than `Summary.ref`.

**Comment by ejgallego** [comment]
@cpitclaudel @siegebell , mutable state is kept in the states by registering it with the [`Summary`](https://github.com/coq/coq/blob/trunk/library/summary.mli) module. Basically, whatever is registered with the summary, will be saved for each state. Thus, we you move to that state (usually by issuing `EditAt`), you get the same info. 

That being said, LtacProf register its data [here](https://github.com/coq/coq/blob/trunk/ltac/profile_ltac.ml#L84), so the treenode will get restored. Of course, it could be tweaked as the users want.

**Comment by JasonGross** [comment]
@ejgallego @gares  What would it take to support parallel workers and merge the profiling data given by multiple workers?

**Comment by ejgallego** [comment]
I'm afraid I don't know the exact details so @gares will have to answer, but I'd bet that is doable without too much hassle.

**Comment by siegebell** [comment]
I've updated the PR to offer a more sensible data structure to the user. (I don't think it is good to hide the total time as the *negation of an arbitrarily-named root attribute*, to have a bunch of dummy attributes at the root, or to specify the name of a tactic separately from its statistics.)

The results of LtacProfResults now looks like:
```xml
<ltacprof total_time="0.2">
  <list>
    <ltacprof_tactic name="foo" total="0.2" self="0.1" num_calls="3" max_total="0.5">
      <list>
        <ltacprof_tactic name="auto" total="0.1" self="0.1" num_calls="10" max_total="0.3">
          <list/>
        </ltacprof_tactic>
        <ltacprof_tactic name="discriminate" total="0.1" self="0.1" num_calls="2" max_total="0.1">
          <list/>
        </ltacprof_tactic>
      </list>
    </ltacprof_tactic>
  </list>
</ltacprof>
```

**Comment by siegebell** [comment]
@JasonGross @ejgallego @cpitclaudel -- What do you think about supporting both *stateful* **and** *stateless* profiling? I believe that the former is ideal for the REPL model and the latter is better for IDEs. The vernacular (and possibly ltac) commands to control profiling make most sense when it is stateful, so I think that behavior has to stay. But if the user issues a command to set/unset/reset profiling from the IDE, there is no clear way to indicate how the profiling results relate to the code in front of them. This will be extremely confusing to the user.

* In *stateful* profiling (the current model), each state has a copy of the profile and `Show Ltac Profile` shows the results for the current state -- but would additionally issue a feedback message with the full structured results demonstrated above. (Alternately, the IDE can look for the feedback message that is already generated and call `LtacProfResults` in response. This would work better if such feedback messages were distinguished from `idtac` output.) `Set Ltac Profiling`, `Unset Ltac Profiling`, and `Reset Ltac Profile` vernac continue to apply to stateful profiling.

* In *stateless* profiling, a single profile is tracked that is not associated with a state and is not rewound when we call `EditAt`. We add commands to the xml-protocol to set/unset/reset only the stateless profile.

* `LtacProfResults` accepts a state_id. When given a nonzero state_id, it reports the structured results for the stateful profiling. When state_id is `0`, it reports the stateless results.

**Comment by ejgallego** [comment]
@siegebell , I have no idea about the cost of each option.Indeed I guess people experienced with Ltac profiling may know what works best.

A side comment thou: for maintenability reasons, it is likely that future versions of Coq generate the xml marshallers automatically form the type descriptions in the `.mli` files. Thus, the xml protocol will get linked to the ML interface (of course the old manual marshaller will be maintained), so I'd try to consider such type data as public API and try to get it reasonably ok.

**Comment by siegebell** [comment]
@ejgallego : This PR also provides "nicer" datatypes for the *.mli file. (LtacProf has its own internal data structures with mutable state, which have never been public.) But does auto-generated marshaling mean that XML attributes should be avoided?

```ocaml
(* from ltac/profile_ltac.mli *)
type ltacprof_entry = {total : float; self : float; num_calls : int; max_total : float}
type ltacprof_tactic = {name: string; statistics : ltacprof_entry; tactics : ltacprof_tactic list }
type ltacprof_results = {total_time : float; tactics : ltacprof_tactic list }
```

**Comment by ejgallego** [comment]
@siegebell , sorry I don't understand what you mean by "XML attributes should be avoided". Automatic marshalling means that it will take:
```ocaml
type ltacprof_entry = {total : float; self : float; num_calls : int; max_total : float}
```
and produce functions:
```ocaml
val ltacprof_entry_of_xml : Xml.t -> ltacprof_entry
val xml_of_ltacprof_entry : ltacprof_entry
```
so you avoid the boilerplate. Of course, many strategies are possible, but the point is that if you change the `type ...` definition, your XML representation may change.

**Comment by ejgallego** [comment]
See https://github.com/janestreet/ppx_xml_conv for an example of such serializer.

**Comment by siegebell** [comment]
Regarding the choice between stateful and stateless profiling, I think that is a decision best left for another PR because it is related to the ongoing debate about how "Set Options" should work for IDEs. (E.g. see the coq-dev thread "Some XML protocol issues" about *synchronous*/stateful and *asynchronous*/stateless options.)

I had provided `LtacProfReset` because there was no way for the IDE to reset the profile without calling `Add` and generating a new state_id, whereas set/unset were already provided by the "Set Options" infrastructure. But until this issue is resolved, I think it is best for the IDE to *never* attempt to set/unset/reset profiling. Because such commands are tied to the current state, they should be explicitly inserted by the user as vernacular commands. Thus I will remove `LtacProfReset`

**Comment by siegebell** [comment]
@ejgallego By  "XML attributes should be avoided", I am wondering whether e.g. `num_calls="3"` should be an attribute of the `ltacprof_tactic` tag or else appear as `<int>3</int>` inside the tag.

**Comment by ejgallego** [comment]
This is up to the serializer I guess. Json ones tend to create an identically named field for such attributes.

**Comment by cpitclaudel** [comment]
+1 for merging this. Let's not get distracted by the (interesting) discussions about semantics of LtacProf; they are orthogonal to this PR.

**Comment by siegebell** [comment]
1. `LtacProfResults` now takes a state_id because profiling is "synchonized" (i.e. attached to a state). Stateid.dummy will provide results for the currently focused state. This command will now "observe" the state; i.e. will wait until the specified state has been processed.
2. `LtacProfReset` has been removed because it's behavior is unintuitive for synchronized profiling. The user should use `Set Ltac Profiling`, `Unset Ltac Profiling`, and `Reset Ltac Profile` commands in their script, and only ask the IDE to display the profile for a given state/sentence.
3. I commented my code.

**Comment by ejgallego** [comment]
Oh that's very interesting @siegebell ; we had a brief chat on coqdev about making a more general API for `Stm.query`. Something like `Stm.with_state`. Indeed, that would fit for your user in `ltacprof_results`.

**Comment by ejgallego** [comment]
In any case marking this bit code in my TODO list, when the STM provides a general "in-state" query mechanism, we should move it to that API.

**Comment by gares** [comment]
Sorry guys, I had no time to look at that.  I will next week and see how to conciliate the need this PR expresses and the freeze.

After a bird eye view, what I would have done is:
- register ltacprof data in the summary
- have explicit Set/Unset to enable/disable profiling in the document (i.e. add)
- have a specific query to print the values as feedback (feedback has a user extensible field)
- add to the query message of the XML protocol the state_id (the Stm is able to run a query at any state).

So that a UI that understands the feedback message can display profiling infos at any state between set/unset profile ltac: the use runs the query there (or the UI does it for him).

Only the last bullet requires changes to Coq, all the rest can go in a plugin (and the first one is a fix if it is not already the case).  I'll check this PR again on Monday, sorry for not answering earlier.

**Comment by ejgallego** [comment]
Sounds great Enrico, do you plan to add a general query function the Stm them?

**Comment by gares** [comment]
Query takes a string, if you VERNAC EXTEND ... "MyQuery" -> my_query () ... END, then you can already run any ML code (my_query) in any state, since Stm.query ~at:id does it.  IIRC the XML protocol does *not* have a state_id in the query data type.

**Comment by ejgallego** [comment]
I see, thanks. I was thinking more of something that would allow [this](https://github.com/coq/coq/pull/209/files#diff-b1cbfdbfc958d51fbc6a77fe784d0f6cR434) to be written as:
```ocaml
Stm.gen_query id Profile_ltac.get_profiling_results
```
Note the minor differences between the current code in `ltac_prof` and what `Stm.query` actually does.
E.

**Comment by ejgallego** [comment]
> IIRC the XML protocol does not have a state_id in the query data type.

This claims differently https://github.com/siegebell/vscoq/blob/master/CoqProtocol.md#command-query (code seems to support)

**Comment by gares** [comment]
My memory sucks.  Yes, it already has a state id!  So I guess the shortest path is to add (if not already present) a syntax to query Ltac Profiling data, register such data in the summary (if not already), and add a case to Feedback carrying the datatype CJ uses (and marshalling to XML as he does).
Would that work for you guys?

I agree your API it something to be added in 8.7.  I don't like Coq's style of zillions of functions.  Only one Stm.query taking ~at:Stateid.t and ~query:[ `Text of string | `ML of (unit -> 'a option)] or something like that.

**Comment by gares** [comment]
Note that if the query for gettin profiling data also prints some  plain text, one can also use it PG or CoqIDE:
```
Set Ltac PRofiling.
tac1.
Show Ltac Profiling.
tac2.
Show Ltac Profiling.
Unset Ltac Profiling.
```
while an IDE that understands the structured feedback mesage can work with just
```
Set Ltac PRofiling.
tac1. tac2.
Unset Ltac Profiling.
```
and give a way get such info by hovering on tac1 or tac2.

**Comment by JasonGross** [comment]
> register ltacprof data in the summary

@gares is 
```coq
let stack =
  let freeze _ = List.map deepcopy_treenode in
  Summary.ref ~freeze [empty_treenode 20] ~name:"LtacProf-stack"
```
enough to register `stack` in the summary?

**Comment by gares** [comment]
Looks good to me.

**Comment by siegebell** [comment]
@gares - I've implemented the changes you suggested. The structured ltac profiling results are now reported as a custom feedback message in response to `Show Ltac Profile` and I have reverted all changes to the xml protocol itself.

However, I am concerned that `Show Ltac Profile` now generates *two* feedbacks and IDEs that handle the structured feedback have have no clear way to distinguish whether a plain-text feedback message is redundant or else needs to be displayed to the user. 

Should I add an option to toggle between generating plain-text feedback messages or structured custom feedback? E.g. `Set Ltac Structured Feedback`.

**Comment by ejgallego** [comment]
Umm, that's an interesting question. I would have expected the profiling message to live in the xmlprotocol (as it was before), given it is backend specific.

It seems to me that profiling data should be explicitly requested, in contrast to feedback which in principle seems to send data that was not explicitly asked for.

At least in SerAPI, no feedback for profiles will be ever generated (other than `Message` objects resulting from Vernaculars). The IDE will have to send an explicit `(Query ((id $ID)) ProfileData)` and sertop will reply with the serialized `ltacprof_results`.

**Comment by siegebell** [comment]
@ejgallego - I agree, all encoding should ideally be done by ide_slave/xml-protocol, SerAPI, etc., so there is something fundamentally wrong with `Custom` accepting xml and expecting a plugin to make the encoding choices.

Since no one currently uses `Custom` (except potentially my patch), now might be the time to find a better way -- but that sounds like a separate PR.

**Comment by cpitclaudel** [comment]
My gut feeling would be to separate `Show Ltac Profile` from the IDE API entirely. As a user if I type `Show Ltac Profile` I'd like to be show the text-based profile, and if I click the profile button I'd like my IDE to start getting all fancy.

**Comment by ejgallego** [comment]
I agree with both of you.

p.s: Enrico mentioned that custom was used by PIDE top.

**Comment by gares** [comment]
@ejgallego I agree, with a serialization machinery queries should have type (... -> 'a) and have that 'a serialized.  But the XML protocol has only 2 ways to give back to the UI: plain text or feedback.  Feedback as a handful hard coded data types you really want to support to understand what Coq Is doing, plus an extensible one.  Yes, XML is an arbitrary choice for the payload, but given we speak XML it is also the most obvious one, and has structure, something the plain text has not (even if the richpp mess added some...)

@siegebell looks good, I'll check the details tomorrow and hopefully merge.

@cpitclaudel I agree, if you type Show Ltac Profile the UI should show you the text output "in the message window".  This works on any UI.  If the UI also understands the feedback message, well it may decide to not show it.  But if the UI proposes a button/hover/whatever to access this piece of info it may drop the plain text and use the structured data instead.

We could have 2 query, one printing text and one feedbacking the structured data, but I don't think it is worth it.

@ejgallego yes, pidetop addds a query to print goals in a structured way and uses the Custom feedback message to print them in a structured way.  But this makes me thik about something you have to solve in serapi too.  eclipse<---->coqtop(pidetop)<--->proofworker.  it is the proof worker that emits the structured feedback for goals, and pidetop forwards it to the UI.  Such machinery is easy for feedback (set_feeder...), you may want to have one for serialized data, but given it is data you get in response to a function call, I quite don't get how you would handle the man in the middle thing. feedback is "broadcasted" in some sense.  I hope I managed to explain the mess.

**Comment by gares** [comment]
Maybe I can be more clear: I have the impression that in serapi you expect an actor to reply, with with feedback messages you don't care which actor will, at some point in time, send you the data.  In other words, can coqtop delegate the job/response to a serapi query to a worker (without keeping track of it, as a NAT server would do if that helps to understand)? Today I can send the task to a worker, the UI does not care who generates the feedback.

**Comment by ejgallego** [comment]
Hi @gares , thanks for the comments!

I guess the point we were making is we saw nothing wrong adding a new entry to the protocol as it was originally done.

Why do you say "the XML protocol has only 2 ways to give back to the UI" ? AFAICS for examples `goals_rty` goes beyond that, doesn't it?

I am not sure I understand what do you mean on the feedback delegation. Regarding goals for example, in SerAPI never generates a "goals" message unless as a reply to a goals query. Goals are never put into Feedback.

Workers only respond to SerAPI so if something needs to be fetched from them, master does it. We forward feedback in the usual way is done today. The SerAPI protocol is in a sense very open. If you need to query a new object, let's say the profiling data, you use the existing query method with a new tag, and answers reply with a new tag too. As I said, Feedback is reserved non-user initiated ("async"?) output. An inconsistency now is error reporting thou, but this is due to Coq: some errors generate feedback, some others not.

**Comment by ejgallego** [comment]
Regarding "actors", yes, when querying for data there is main "actor" replying: master, which provides a consistent view over such data.

**Comment by ejgallego** [comment]
To clarify for instance for the LtacProf case, we'd like SerAPI to operate like this:

1. The IDE receives a `Completed/Processed` feedback for a particular `id : Stateid.t`.
2. The IDE then chooses whether to fetch the `LtacProf` data with a `Query` or not.
3. It could happen that the `id` becomes cancelled between the feedback and the query, then SerAPI will return an empty stream. The IDE may not receive now a `Cancelled` event between the `Query` and the answer, however, if we relax this restriction then the IDE may need to learn to ignore streams for a cancelled id.

Does it make sense?

**Comment by cpitclaudel** [comment]
> It could happen that the id becomes cancelled between the feedback and the query, then SerAPI will return an empty stream. The IDE may not receive now a Cancelled event between the Query and the answer, however, if we relax this restriction then the IDE may need to learn to ignore streams for a cancelled id.

Is anything preventing the following though?

0. SerAPI sends cancelled
1. IDE issues query
2. IDE receives "cancelled"
3. SerAPI receives query
4. SerAPI sends empty response
5. IDE receives empty response 

(I don't think that's a problem btw; the IDE will just never receive an empty response for a non-cancelled state).

**Comment by ejgallego** [comment]
This is allowed for now. This behavior should be similar as requesting profile data for a state that has no profile data; (or goals).

Of course we can tweak it.

**Comment by gares** [comment]
@ejgallego what is wrong with "a new XML message" is that you can't conceive a protocol where any new request/response needs a new message.  Otherwise the protocol changes every time you want to print something.  Maybe I can compare it with ISO/OSI: the XML protocol is TCP, you don't change TCP, you change the upper layer.

You don't see this in serapi, since you are closer to RPC.  The XML protocol requires a lot of boilerplate (in Coq) to add a message.  So unless it is really needed you don't do it.  What do you need to do in order to support a new query in serapi?

@siegebell merges (but squashed) since the history was too complex for the little change.

**Comment by gares** [comment]
@ejgallego goals_rty is there because it is really needed.  There are also other messages there for no good reason (they were there before, no energy to remove them).  hints, cases are examples.  I'm guilty for print_ast that should just be a query+feedback, also annotate I'm afraid.

The way I see the XML protocol is: some messages to communicate the document and the processing status + query/feedback for getting more data.  Given that the language of query and feedback is extensible, the protocol needs not be changed every time one adds a functionality.

**Comment by ejgallego** [comment]
Hi @gares , I'm not sure what you mean, many protocols and languages are extensible, including TCP (see RFC7323 for example).

In fact, `Feedback.feedback` already is extensible by virtue of the Custom message, even if in a poor way as you cannot have a tag for the custom message.

My point here is that the public API is `type ltacprof_results`. How to encode such data structure is the responsibility of the particular protocol in use. Previously to this latest patch, the IDE/XML protocol-specific encoding was self-contained in `Xmlprotocol`, now is not. Apart from that, having an effectful interface for something as simple as querying the data seems a bit overkill.

Imagine the scenario of an independent `CoqIDE` like the one in the split tree. Now, if you want to tweak the XML realization, you need to patch Coq. Previously you didn't.

IMHO implementation details are a different issue. In SerAPI you have object tags, you just add a new tag and you get a new `Query`/`Answer` types. For this particular example this is what it took: ejgallego/coq-serapi@5da82eb 

By the way, IMHO Profile_ltac should export the corresponding standard `pr_$data : $data -> Pp.std_ppcmds` functions .

**Comment by ejgallego** [comment]
> Given that the language of query and feedback is extensible, the protocol needs not be changed every time one adds a functionality.

Oh well `Feedback` is part of the protocol this way, isn't it? Another thing I don't understand with this patch is that now you require a new Vernac for every new query ?

**Comment by gares** [comment]
Well, it has always been like that (in REPL).  And I agree with you, query should take any function, not just the ones with a concrete syntax.

Your point about modifying Coq to change the XML representation of profile_data is, IMO, moot. The XML realization should be isomorphic to the data type.  You change the data, you change the serialization.  In serapi the scenario you depict is never going to happen: a data type that you want to tune its serialization without changing the datatype itself.  I agree the current way of doing things make the scenario possible in Coq, since serialization to XML is not automatic, but should really be considered as such.  I'd love to replace all to/of_xml code to generated code.  

Having the ppx code hardcoded only in the toploop plugin is problematic I'm afraid. You need to fix the toploop as in the commit you pointed out every time a new message is needed.  But Coq is itself extensible via plugins, that define datatypes and may want to print them.  Also UI come with extension mechanisms.  Now, if you have a generic query/feedback mechanism you don't need to touch the toploop nor the protocol in order to get more info out of the prover.  Eg Coqoon added a query for structured goals without changing Coq nor the protocol.  If I understand correctly, this is not currently possible in serapi.
In other words, I guess you need to stick into the Custom feedback message an object with methods to serialize it to sexp/xml/whtever.

Maybe I can make an example to clarify.  Imagine the quick-check plugin works like that: it registers itself to a hook "a new thm stated" and if it finds a counter example it sends a feedback.  A UI understanding such feedback can display it.  quick-check is a Coq plugin, the UI uses the XML protocol.  No change is needed to the XML protocol, no single line change in Coq.
Can you do that?  Do you want to support such scenario?

**Comment by ejgallego** [comment]
Hi Enrico, I just meant that there may be other choices of serializers and serialization strategies IMVHO should be left to the client. Even if serialization is automatic it can happen in many ways (think of proto-buffers, etc...)

> Having the ppx code hardcoded only in the toploop plugin is problematic I'm afraid.

That's a choice mainly due to keep things simple in the experimental prototype. If there comes an use case where plugins want to extend SerAPI it is possible to provide registration functions for them. The Sexp serialiazers can be made open.

>  Eg Coqoon added a query for structured goals without changing Coq nor the protocol. If I understand correctly, this is not currently possible in serapi.

I'm not sure I understand. Coqoon had to write a whole new toplevel to do this! The custom message is supported in SerAPI in the same way than any other feedback.

Precisely the point of SerAPI is to avoid what had to be done in Coqoon: you never need to use a custom feedback type for core Coq datatypes. If you are a plugin and you are providing a new datatype, I'd be happy to provide an open serializer so the plugin can hook. Any plugins needing that, maybe ssreflect?

> Imagine the quick-check plugin works like that: it registers itself to a hook "a new thm stated" and if it finds a counter example it sends a feedback. A UI understanding such feedback can display it. quick-check is a Coq plugin, the UI uses the XML protocol. No change is needed to the XML protocol, no single line change in Coq. Can you do that? Do you want to support such scenario?

That's a great example, thank you! Currently, SerAPI clients would have to use this route, which is supported. However, in the SerAPI model, the plugin would register a new tag and serializer. Then, Query, Print, etc.. would gain knowledge of the tag and dispatch properly.

The protocol wouldn't change, but extended. The key difference is that counterexample objects would only be sent back to the client if requested with a query first in an stream datatype.

Providing plugin registration is something I was planning to do more like in a 2.0 version but I'd be happy to code it if somebody points me to a suitable plugin.

**Comment by gares** [comment]
So, in your view, Feedback is only for spontaneously generated metadata.  While queries are on demand. Am I right?

I guess we are mixing the semantics of the word "feedback" with what it does: it is an asynchronous communication of data from the prover to the UI.  It's most natural use is feedback like OK/Error. But  I guess it is the behavior you want for your queries' output.

Which type would you give to Stm.query ?  Wouldn't the "output" of such API behave exactly as a feedback (attached to an handle, asynchronous...)?

**Comment by ejgallego** [comment]
Indeed I took the view:
- Streams: response to IDE-initiated Searchs or Queries.
- Feedback: Non-user initiated information relayed by Coq.

Both are "asynchronous".

Indeed, recall that indeed what we do in SerAPI is 100% and evolution or extension of the current protocol. In particular, `Stm.query` is exactly SerAPI `Query`, extended to be over tagged objects with filters, and choice of output format (*) In particular this last point made me to introduce Streams and not reuse Feedback to return query results.

Does it make sense?

I should update the notes in the repos, I also have two CEPs in the works so we can discuss a bit more structuredly.

[*] (_note that the current implementation is not finished_)

**Comment by ejgallego** [comment]
Another reason not to overload the feedback too much (apart from limiting, etc...) is that sometimes query preparation may be very expensive in terms of time!

**Comment by gares** [comment]
So I guess both carry the state_id they refer to.  Ok, so it is just that feedback is 1 message, and streams are a bunch of related feedback messages.  Did I get it?  the one is spontaneous, the other is not, but as messages they are "effects".  Query still returns unit, right?

**Comment by ejgallego** [comment]
Conceptually (and internally) query in SerAPI is of type ` query : filter_predicate \to coq_object stream`

Streams don't carry the `state_id` for now; but it could be added if useful.

**Comment by gares** [comment]
Actually feedback messages already have a route_id, that Coqoon uses to dispatch messages to appropriate windows.

**Comment by gares** [comment]
The idea being that a query has, on the UI side, a receiver and the route is used to sort things out on the UI side.  IIRC, Stm.query takes the route in input, and attaches it to all feedbacks coming from the query.

**Comment by siegebell** [comment]
@gares I was wondering what the heck `route` was for.

**Comment by siegebell** [comment]
@gares The xml-protocol doesn't seem to accept `route`. And I don't think this should be limited to queries -- *all* calls could accept a routing parameter.

**Comment by gares** [comment]
You are right, it could.
It was added to Query because we had a use case, if the XML does not have it I can surely add it (sine the ML api has it).
Still, I quite don't see how it could serve a purpose in add or edit_at.  Do you have an example in mind?

**Comment by siegebell** [comment]
@gares - you're right, I can't think of any particular need, either. Anyhow, having xml-protocol support for routing IDs on queries would be very helpful!
